tla-pp/src/main/scala/at/forsyte/apalache/tla/pp/temporal/utils.scala

```diff
@@ -0,0 +1,86 @@
+package at.forsyte.apalache.tla.pp.temporal
+
+import at.forsyte.apalache.tla.lir._
+import at.forsyte.apalache.tla.typecomp._
+import at.forsyte.apalache.tla.lir.transformations.TransformationTracker
+import at.forsyte.apalache.tla.lir.transformations.standard.Flatten
+import at.forsyte.apalache.tla.lir.TypedPredefs.TypeTagAsTlaType1
+
+object ScopedBuilderExtensions {
+  implicit class ScopedBuilderExtension(val builder: ScopedBuilder) {
+    def primeVar(varDecl: TlaVarDecl): TBuilderInstruction = {
+      builder.prime(declAsNameEx(varDecl))
+    }
+
+    def declAsNameEx(decl: TlaDecl): TBuilderInstruction = {
+      builder.name(decl.name, decl.typeTag.asTlaType1())
+    }
+  }
+}
+
+/**
+ * A convenience class storing a module, together with the init, next and loopOK predicates of that module. Useful to
+ * avoid re-finding these predicates.
+ */
+class ModWithPreds(
+    val module: TlaModule,
+    val init: TlaOperDecl,
+    val next: TlaOperDecl,
+    val loopOK: TlaOperDecl) {
+
+  def setPredicates(newInit: TlaOperDecl, newNext: TlaOperDecl, newLoopOK: TlaOperDecl): ModWithPreds = {
+    val newDeclarations = module.declarations.map(decl =>
+      decl.name match {
+        case init.name   => newInit
+        case next.name   => newNext
+        case loopOK.name => newLoopOK
+        case _           => decl
+      })
+    val newModule = new TlaModule(module.name, newDeclarations)
+    new ModWithPreds(newModule, newInit, newNext, newLoopOK)
+  }
+
+  def setModule(newModule: TlaModule): ModWithPreds = {
+    new ModWithPreds(newModule, init, next, loopOK)
+  }
+
+  def prependDecl(decl: TlaDecl): ModWithPreds = {
+    val newDecls = decl +: module.declarations
+    val newModule = new TlaModule(module.name, newDecls)
+    setModule(newModule)
+  }
+
+  /**
+   * Replaces all instances of oldDecl with newDecl
+   */
+  def replaceDeclInMod(oldDecl: TlaDecl, newDecl: TlaDecl): ModWithPreds = {
+    val newDeclarations = module.declarations.map(decl => if (decl.name == oldDecl.name) newDecl else decl)
+    val newModule = new TlaModule(module.name, newDeclarations)
+    new ModWithPreds(newModule, init, next, loopOK)
+  }
+}
+
+package object utils {
+  val builder: ScopedBuilder = new ScopedBuilder()
+}
+
+object DeclUtils {
+  import at.forsyte.apalache.tla.pp.temporal.utils.builder
+
+  /**
+   * Takes decl, ex and returns newDecl with the same name as decl, with its body extended like this: newDecl == decl /\
+   * ex
+   */
+  def conjunctExToOperDecl(ex: TlaEx, decl: TlaOperDecl, tracker: TransformationTracker): TlaOperDecl = {
```

Comment:
```suggestion
  def andInDecl(ex: TlaEx, decl: TlaOperDecl, tracker: TransformationTracker): TlaOperDecl = {
```

In general, I keep longer-living objects on the left-hand side. So I would prefer this order of the arguments: `tracker`, `decl`, and `ex`

Reply:
Interesting. My intuition is to have the "primary actors" on the left-hand side, and "background actors" on the right-hand side.
E.g. `AdditionThatAlsoLogs(summand1, summand2, logger)`
If you prefer tracker, decl, ex nonetheless, I'll adjust :)

Reply:
I agree with Philip on this; I'd find it confusing if `tracker` was the first parameter here.